request.go

```diff
@@ -1253,11 +1253,18 @@ func (r *Request) TraceInfo() TraceInfo {
 
 	// Calculate the total time accordingly,
 	// when connection is reused
+	var requestStartTime time.Time
 	if ct.gotConnInfo.Reused {
-		ti.TotalTime = ct.endTime.Sub(ct.getConn)
+		requestStartTime = ct.getConn
 	} else {
-		ti.TotalTime = ct.endTime.Sub(ct.dnsStart)
+		requestStartTime = ct.dnsStart
 	}
+	// DNS start and get conn time may be zero if the request is invalid.
+	// See issue #1016.
+	if requestStartTime.IsZero() {
+		requestStartTime = r.Time
+	}
+	ti.TotalTime = ct.endTime.Sub(requestStartTime)
```

Comment:
Maybe it could be rewritten like this then

```go
          // DNS start and get conn time may be zero if the request is invalid, so default one is the request time
          // See issue #1016. 
          requestStartTime := r.Time

          if !ct.dnsStart.IsZero() { 
                 requestStartTime = ct.dnsStart 
          }

          if ct.gotConnInfo.Reused && !ct.getConn.IsZero() { 
                 requestStartTime = ct.getConn 
          } 

         ti.TotalTime = ct.endTime.Sub(requestStartTime)
```

This way if only one of is no zero, it works

Reply:
@ccoVeille Thank you! I changed the code style like this.